src/app/modules/spcp/spcp.oidc.service.ts

```diff
@@ -0,0 +1,511 @@
+      action: 'parseState',
+      state,
+      authType,
+    }
+    const payloads = state.split('-')
+    const formId = extractFormId(payloads[0])
+    if ((payloads.length !== 2 && payloads.length !== 3) || !formId) {
+      logger.error({
+        message: 'state incorrectly formatted',
+        meta: logMeta,
+      })
+      return err(new InvalidStateError())
+    }
+
+    const rememberMe = payloads[1] === 'true'
+    const encodedQuery = payloads.length === 3 ? payloads[2] : ''
+    let decodedQuery = ''
+
+    try {
+      decodedQuery = encodedQuery
+        ? `?${Buffer.from(encodedQuery, 'base64').toString('utf8')}`
+        : ''
+    } catch (e) {
+      logger.error({
+        message: 'Unable to decode encodedQuery',
+        meta: {
+          encodedQuery,
+          ...logMeta,
+        },
+        error: e,
+      })
+      return err(new InvalidStateError())
+    }
+
+    const destination = `${payloads[0]}${decodedQuery}`
+
+    let cookieDuration
+    if (authType === FormAuthType.CP) {
+      cookieDuration = this.#spcpOidcProps.cpCookieMaxAge
+    } else {
+      cookieDuration = rememberMe
+        ? this.#spcpOidcProps.spCookieMaxAgePreserved
+        : this.#spcpOidcProps.spCookieMaxAge
+    }
+
+    return ok({
+      formId,
+      destination,
+      rememberMe,
+      cookieDuration,
+    })
+  }
+
+  /**
+   * Creates a JWT with a payload of SP user data.
+   * @param payload Information to add to JWT
+   * @param authType FormAuthType.SP | FormAuthType.CP
+   * @param cookieDuration Cookie validity duration
+   * @return okAsync(The JWT in a string)
+   * @return errAsync(CreateJwtError)
+   */
+  createJWT(
+    payload: JwtPayload,
+    cookieDuration: number,
+    authType: FormAuthType.SP | FormAuthType.CP,
+  ): ResultAsync<string, CreateJwtError> {
+    const logMeta = {
+      action: 'createJWT',
+      authType,
+    }
+
+    const client = this.getClient(authType)
+
+    return ResultAsync.fromPromise(
+      client.createJWT(
+        payload,
+        `${cookieDuration / 1000}s`,
+        // Sets "exp" (Expiration Time) Claim value on the JWT Claims Set. When number is passed that is used as a value
+        // When string is passed it is resolved to a time span and added to the current timestamp.
+      ),
+      (error) => {
+        logger.error({
+          message: 'Failed to create JWT',
+          meta: {
+            ...logMeta,
+          },
+          error,
+        })
+        return new CreateJwtError()
+      },
+    )
+  }
+
+  /**
+   * Creates a payload of SP/CP user data based on attributes
+   * @param attributes user data returned by SP/CP from client
+   * @param rememberMe Whether to enable longer duration for SingPass cookies
+   * @param authType FormAuthType.SP | FormAuthType.CP
+   * @return The payload
+   */
+  createJWTPayload(
+    attributes: string | ExtractedCorppassNDIPayload,
+    rememberMe: boolean,
+    authType: FormAuthType.SP | FormAuthType.CP,
```

Comment:
Having most (all?) methods take an `authType` and then do conditional checks doesn't feel like the cleanest OO 🤔 I wonder if `spcp.oidc.service` could export 2 services `spService` and `cPService` instead, while having an internal base class that holds the common structure.

I make the comment after seeing that the client of the service typically do their checks on the `authType` first, so it feels like we could avoid having to do the checks at multiple level. 

Admittedly, I haven't thought very deeply about how hard it'd be to refactor 😓 , but I'd be keen to try if you want to pair.

Reply:
I think the refactor idea applies to the OIDC client class tooSince there are the subclasses by types already, the base class should not need to be aware of the authType. The base classes themselves should represent this awareness.

I'm going to give it a quick shot at refactoring from the 2 prior comments today for POC (also to have a sense if it's harder than I assumed here). Will let you know when done.